### PR TITLE
Fixes all the problems from the last PR I think?

### PR DESCRIPTION
**Pull request by psygnisfive**

### CODE REVIEW

File: plutus-core-spec/figures/PlutusCoreGrammar.tex [hunk 1]
```diff
@@ -5,45 +5,47 @@
...
         \textrm{Type}             & A,B,C  & ::= & \alpha                     & \textrm{type variable}\\
+                                  &        &     & s                          & \textrm{size}\\
                                   &        &     & \funT{A}{B}                & \textrm{function type}\\
                                   &        &     & \allT{\alpha}{K}{A}        & \textrm{polymorphic type}\\
                                   &        &     & \fixT{\alpha}{A}           & \textrm{fixed point type}\\
                                   &        &     & \lamT{\alpha}{K}{A}        & \textrm{$\lambda$ abstraction}\\
                                   &        &     & \appT{A}{B}                & \textrm{function application}\\
-                                  &        &     & \conT{tcn}                 & \textrm{builtin type}\\
-        \textrm{Type Value}       & R,S,T  & ::= & \funT{S}{T}                & \textrm{function type}\\
+                                  &        &     & \conT{tcn}{A}          & \textrm{type constant}\\
```
Comment (effectfully) [inline plutus-core-spec/figures/PlutusCoreGrammar.tex]: From the previous thread:

> Oh! I've reorgged the way builtins and constants work, so now type constants are just a name and a size (the A arg). I would have added a signature component for them too but all type constants take size args only. If you think having a signature system would be better, let me know.

I do not have any opinion here right now (that's a "being ad hoc vs overcomplicating" dilemma), but if we're going to generalize built-ins (that `dyn` type variable and stuff), this might be not that relevant now.

File: plutus-core-spec/figures/PlutusCoreReduction.tex [hunk 2]
```diff
@@ -60,6 +61,7 @@
                        %&   &     & \inLamLeftFrame{x}{M}                   & \textrm{$\lambda$}\\
                        &   &     & \inAppLeftFrame{M}                      & \textrm{left app}\\
                        &   &     & \inAppRightFrame{V}                     & \textrm{right app}\\
+                       &   &     & \inBuiltin{bn}{A*}{V^*}{\_}{M^*}        & \textrm{builtin}\\
```
Comment (effectfully) [inline plutus-core-spec/figures/PlutusCoreReduction.tex]: Since here we have n-ary type instantiations and applications, would it make sense to have a hole for each of the term arguments? I.e. instead of one `_`, we could have an entire `_^*` and fill them all at the same time. I think it would simplify both the spec and the implementation. Or does this not make sense?

File: plutus-core-spec/figures/PlutusCoreReduction.tex [hunk 3]
```diff
@@ -144,27 +143,27 @@
...
+        &\cksteps{\ckforward{s}{\con{cn}}}{\ckbackward{s}{\con{cn}}}\\
         &\cksteps{\ckforward{s}{\abs{\alpha}{K}{M}}}{\ckbackward{s}{\abs{\alpha}{K}{M}}}\\
         &\cksteps{\ckforward{s}{\inst{M}{A}}}{\ckforward{s, \inInstLeftFrame{A}}{M}}\\
         &\cksteps{\ckforward{s}{\wrap{\alpha}{A}{M}}}{\ckforward{s, \inWrapRightFrame{\alpha}{A}}{M}}\\
         &\cksteps{\ckforward{s}{\unwrap{M}}}{\ckforward{s, \inUnwrapFrame{}}{M}}\\
         &\cksteps{\ckforward{s}{\lam{x}{A}{M}}}{\ckbackward{s}{\lam{x}{A}{M}}}\\
         &\cksteps{\ckforward{s}{\app{M}{N}}}{\ckforward{s, \inAppLeftFrame{N}}{M}}\\
-        &\cksteps{\ckforward{s}{\con{cn}}}{\ckbackward{s}{\con{cn}}}\\
-        &\cksteps{\ckforward{s}{\error{S}}}{\ckerror{}}\\
+        &\cksteps{\ckforward{s}{\builtin{bn}{\repetition{A}}{}}}{U} \qquad \textit{$bn$ computes on $\repetition{A}$ to $U$ according to \ref{fig:Plutus_core_builtins}}\\
```
Comment (effectfully) [inline plutus-core-spec/figures/PlutusCoreReduction.tex]: I'm not sure how to read this. A built-in is only applied to a spice of types, but is able to reduce somehow?

File: plutus-core-spec/figures/PlutusCoreReduction.tex [hunk 4]
```diff
@@ -144,27 +143,27 @@
 
 \begin{figure*}[t]
     \judgmentdef{\(\cksteps{\sigma}{\sigma'}\)}{Machine state $\sigma$ transitions in one step to machine state $\sigma'$}
-    
+
     \begin{align*}
+        &\cksteps{\ckforward{s}{\con{cn}}}{\ckbackward{s}{\con{cn}}}\\
         &\cksteps{\ckforward{s}{\abs{\alpha}{K}{M}}}{\ckbackward{s}{\abs{\alpha}{K}{M}}}\\
```
Comment (effectfully) [inline plutus-core-spec/figures/PlutusCoreReduction.tex]: Replace `M` by `V` due to the value restriction?

File: plutus-core-spec/figures/PlutusCoreReduction.tex [hunk 5]
```diff
@@ -144,27 +143,27 @@
...
         &\cksteps{\ckforward{s}{\wrap{\alpha}{A}{M}}}{\ckforward{s, \inWrapRightFrame{\alpha}{A}}{M}}\\
         &\cksteps{\ckforward{s}{\unwrap{M}}}{\ckforward{s, \inUnwrapFrame{}}{M}}\\
         &\cksteps{\ckforward{s}{\lam{x}{A}{M}}}{\ckbackward{s}{\lam{x}{A}{M}}}\\
         &\cksteps{\ckforward{s}{\app{M}{N}}}{\ckforward{s, \inAppLeftFrame{N}}{M}}\\
-        &\cksteps{\ckforward{s}{\con{cn}}}{\ckbackward{s}{\con{cn}}}\\
-        &\cksteps{\ckforward{s}{\error{S}}}{\ckerror{}}\\
+        &\cksteps{\ckforward{s}{\builtin{bn}{\repetition{A}}{}}}{U} \qquad \textit{$bn$ computes on $\repetition{A}$ to $U$ according to \ref{fig:Plutus_core_builtins}}\\
+        &\cksteps{\ckforward{s}{\builtin{bn}{\repetition{A}}{M \repetition{M}}}}{\ckforward{s, \inBuiltin{bn}{\repetition{A}}{}{\_}{\repetition{M}}}{M}}\\
+        &\cksteps{\ckforward{s}{\error{A}}}{\ckerror{}}\\
         &\cksteps{\ckbackward{s, \inInstLeftFrame{A}}{\abs{\alpha}{K}{M}}}{\ckforward{s}{M}}\\
```
Comment (effectfully) [inline plutus-core-spec/figures/PlutusCoreReduction.tex]: Replace M by V due to the value restriction?

File: plutus-core-spec/figures/PlutusCoreTypeCorrectness.tex [hunk 6]
```diff
@@ -132,72 +134,77 @@
...
-        \AxiomC{$Q = \mathcal{E}\{\fixT{\alpha}{S}\}$}
-        \AxiomC{$\Gamma \vdash M : \mathcal{E}\{[\fixT{\alpha}{S}/\alpha]S\}$}
-        \RightLabel{wrap}
-        \TrinaryInfC{\(\hypJ{\Gamma}{\istermJ{\wrap{\alpha}{S}{M}}{Q}}\)}
+        \AxiomC{\(\hypJ{\Gamma}{\istypeJ{A}{\typeK{}}}\)}
+        \RightLabel{error}
+        \UnaryInfC{\(\hypJ{\Gamma}{\istermJ{\error{A}}{\normalform{A}}}\)}
     \end{prooftree}
 
-Where $\mathcal{E}$ is an elimination context $\mathcal{E} ::= \bullet \mid [\mathcal{E} \,\, A]$, $A$ a type, such that $\mathcal{E}\{Q\}$ denotes the type expression obtained by replacing the $\bullet$ in $\mathcal{E}$ with $Q$. That is, $[\bullet\,\, A]\{B\} = [B\,\,A]$, and so on.
```
Comment (effectfully) [inline plutus-core-spec/figures/PlutusCoreTypeCorrectness.tex]: What's the status of elimination contexts, why was this deleted, etc?

File: plutus-core-spec/main.tex [hunk 7]
```diff
@@ -699,7 +703,7 @@
 %
 % paper title
 % can use linebreaks \\ within to get better formatting as desired
-\title{Formal Specification of\\the Plutus Core Language v1.0 (RC5.3)}
+\title{Formal Specification of\\the Plutus Core Language v1.0 (RC5.4)}
```
Comment (effectfully) [inline plutus-core-spec/main.tex]: I've seen a RC5.4 which was different from this one. Can we jump straight to 5.5?

File: plutus-core-spec/main.tex [hunk 8]
```diff
@@ -798,19 +802,43 @@
...
+encodings since they may require linear rather than constant time to
+access the components of a data structure. Primitive types, such as
+Integer and Bytestring, are indexed by their size, which allows the
+cost (in gas) of operations such as addition to be determined at
+compile time. In contrast, IELE uses unbounded integers, which never
+overflow, but require that the gas used is calculated at run time. We
+use \Fomega{} (as opposed to F) because it supports types indexed by
+size and parameterised types (such as ``List of A'').
+
+Plutus Core also notable lacks obvious built-in support for recursion,
```
Comment (effectfully) [inline plutus-core-spec/main.tex]: *notably

File: plutus-core-spec/main.tex [hunk 9]
```diff
@@ -798,19 +802,43 @@
...
+either general or otherwise. The reason for this is simply that in
+the presence of type left fixed points, it's possible to give a valid
```
Comment (effectfully) [inline plutus-core-spec/main.tex]: "left"?

File: plutus-core-spec/figures/PlutusCoreBuiltins.tex [hunk 10]
```diff
@@ -4,138 +4,142 @@
...
         $unitval$ & \(\abs{\alpha}{\typeK{}}{\lam{x}{\alpha}{x}}\)\\\\
-        
-        $boolean$ & \(\forall \alpha :: \star.\ \alpha \to \alpha \to \alpha\)\\\\
-        
-        $true$ & \(\abs{\alpha}{\typeK{}}{\lam{t}{\alpha}{\lam{f}{\alpha}{\app{t}{unitval}}}}\)\\\\
-        
-        $false$ & \(\abs{\alpha}{\typeK{}}{\lam{t}{\alpha}{\lam{f}{\alpha}{\app{f}{unitval}}}}\)\\\\
-        
+
+        $boolean$ & \(\forall \alpha :: \star.\ (unit \to \alpha) \to (unit \to \alpha) \to \alpha\)\\\\
```
[the remark this answers is not in the document]
Comment (psygnisfive) [reply plutus-core-spec/figures/PlutusCoreBuiltins.tex]: D'oh. Fixed in my current version (next PR).

File: plutus-core-spec/figures/PlutusCoreGrammar.tex (hunk 1 above)
Comment (psygnisfive) [reply plutus-core-spec/figures/PlutusCoreGrammar.tex]: I'm not sure what you mean?

File: plutus-core-spec/figures/PlutusCoreReduction.tex (hunk 2 above)
Comment (psygnisfive) [reply plutus-core-spec/figures/PlutusCoreReduction.tex]: It doesn't make sense because this is the mechanism for  imposing reduction order, which is not parallel.

File: plutus-core-spec/figures/PlutusCoreGrammar.tex (hunk 1 above)
Comment (effectfully) [reply plutus-core-spec/figures/PlutusCoreGrammar.tex]: I mean, there is a dilemma, but it's probably not important now. I just wanted to make a note which can be of more relevance later.

**Comment by psygnisfive** [comment]
# Builtins

Replaces the unified builtin approach to a split builtin-constant approach, and adds signatures.

Term builtins are now for function-like things only, and have signatures of the form `[a :: K, ...](A, ...)B` which is roughly the same as `forall a :: K, ... A -> ... -> B`.

Constant builtins have been removed to a separate term constant construct, which captures the idea of these being value forms not computing forms. Term constants also have a signature, consisting of a pair of a type constant name and a size. All term constants synthesize types of the form `(conT tcn s)`, so we do not have a general type as the constant's signature.

Type builtins are now called type constants, to align them with term constants as things which have no computational content. They have no signature, and instead all take size arguments by necessity.



# Booleans

Booleans were changed to be lazy, using thunks. This change will be reverted shortly, pursuant to conversations about where this laziness should exist. The consensus on this is that boolean values themselves are strict, but to get lazy if-then-else, we use strict booleans to eagerly select the delayed branches:

  ifThenElse :: (forall a. a -> a -> a) -> forall a. (() -> a) -> (() -> a) -> a
  ifThenElse b t f = b t f ()



# btime

The `btime` parameter to the normalizer has been removed, since we're no longer allowing explicit reference to block time.



# Type Values in Terms

All terms that mention types now permit the use of non-normal types instead of restricting to type values as before. The constraint that on-chain terms should never include non-normal types was consensed as being something imposed by the chain itself, not the language.



# Abstraction Value Restriction

The value restriction on the body of a type abstract `(abs a K ...)` has been reintroduced pursuant to conversations MPJ had, reversing previous changes intended to make formalization easier.



# InFix Reintroduced

Previously, the scopes of type binders had not required normalization, but this is problematic for equality testing, and so they now do. The `Fix` construct was accidentally left out of spec updates for this, so now the appropriate reduction context has been introduced to repair this.



# Fix Scope Kind

The kind of `Fix`'s scope is constrained to be `(typeK)`. This will be changed later when we update `Fix` overall pursuant to Chad's recommendations, but these changes haven't been made anywhere yet in the spec.



# Lambda Argument Type

The argument to a lambda should always be a type of kind `(type)`, but this check was accidentally left out of previous versions of the spec, leading to potentially nonsensical well-typed terms. This has now been fixed.

**Comment by michaelpj** [comment]
> This will be changed later when we update Fix overall pursuant to Chad's recommendations, but these changes haven't been made anywhere yet in the spec.

I thought we'd already merged the elimination contexts stuff in?

**Comment by michaelpj** [comment]
> Booleans were changed to be lazy, using thunks.

No, we changed them to be strict. And I already made the changes to the spec for that, I don't think you need to do anything (indeed, you've undone some of them!). Have a look at the history.

File: plutus-core-spec/figures/PlutusCoreReduction.tex (hunk 2 above)
Comment (psygnisfive) [reply plutus-core-spec/figures/PlutusCoreReduction.tex]: Does it? It ought to be something along the lines of

```
normalize env (Builtin bn _ ms) = normalizeBuiltin bn  <$> mapM (normalize env) ms
```

File: plutus-core-spec/figures/PlutusCoreReduction.tex (hunk 3 above)
Comment (psygnisfive) [reply plutus-core-spec/figures/PlutusCoreReduction.tex]: A builtin applied to an empty spine of terms, so it can compute immediately because all of its arguments have finished computing, whereas a builtin applied to a non-empty spine of terms must begin to compute its spine. I could separate out the transition system into two separate judgments (one for term computation and one for spin computation) if people think that's clearer?

File: plutus-core-spec/figures/PlutusCoreReduction.tex (hunk 4 above)
Comment (psygnisfive) [reply plutus-core-spec/figures/PlutusCoreReduction.tex]: Done.

File: plutus-core-spec/figures/PlutusCoreReduction.tex (hunk 5 above)
Comment (psygnisfive) [reply plutus-core-spec/figures/PlutusCoreReduction.tex]: Done.

File: plutus-core-spec/main.tex (hunk 7 above)
Comment (psygnisfive) [reply plutus-core-spec/main.tex]: Done.

File: plutus-core-spec/main.tex (hunk 9 above)
Comment (psygnisfive) [reply plutus-core-spec/main.tex]: s/ left/-level/

Done.

File: plutus-core-spec/main.tex (hunk 8 above)
Comment (psygnisfive) [reply plutus-core-spec/main.tex]: Done.

File: plutus-core-spec/figures/PlutusCoreTypeCorrectness.tex (hunk 6 above)
Comment (psygnisfive) [reply plutus-core-spec/figures/PlutusCoreTypeCorrectness.tex]: Blergh. Oversight on my part somehow. Fixed.

File: plutus-core-spec/figures/PlutusCoreReduction.tex (hunk 2 above)
Comment (effectfully) [reply plutus-core-spec/figures/PlutusCoreReduction.tex]: That's what I thought of when I asked about "n-ary holes". But the implementations of the CK and CEK machines seem to become more complicated. I'm not entirely sure.

File: plutus-core-spec/figures/PlutusCoreReduction.tex (hunk 3 above)
Comment (effectfully) [reply plutus-core-spec/figures/PlutusCoreReduction.tex]: > because all of its arguments have finished computing

But a built-in does need its arguments in order to compute? And they are just not there.

However, I see how this can work for built-ins that do not receive any arguments.

> I could separate out the transition system into two separate judgments (one for term computation and one for spin computation) if people think that's clearer?

That could be clearer, but I need to play with this and let's not block the PR on this.

**Comment by michaelpj** [comment]
You deleted a lot of files in this PR, was that deliberate?

**Comment by michaelpj** [comment]
Well, the spec builds, so I guess it must have been fine.

The CI notifications are stuck again, but I checked hydra and it's all green, so merging.
